File: frontend/js/views/accrued_interest.js
// Bond accrued interest + dirty price — clean price plus the interest accrued
// since the last coupon, via /calc/accrued-interest. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const num = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 4 });
const VIEW = 'accrued-interest';
let lastReport = null;
let lastBody = null;

export async function renderAccruedInterest(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.accrued.h1.title">// ACCRUED INTEREST</span></h1>
        <p class="muted small" data-i18n="view.accrued.hint.intro">
            Between coupon dates a bond buyer owes the seller the interest accrued since the last
            coupon. The quoted clean price excludes it; the dirty price you actually pay adds it
            back. Pick the day-count convention (30/360 for corporates/munis, Actual/Actual for
            Treasuries). Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.accrued.h2.inputs">The bond</h2>
            <form id="accrued-form" class="inline-form">
                <label><span data-i18n="view.accrued.label.face">Face value ($)</span>
                    <input type="number" step="0.01" min="0" name="face_value" value="1000" required></label>
                <label><span data-i18n="view.accrued.label.coupon">Coupon rate (%)</span>
                    <input type="number" step="0.001" min="0" name="coupon_rate_pct" value="6" required></label>
                <label><span data-i18n="view.accrued.label.freq">Coupons / year</span>
                    <select name="frequency">
                        <option value="1" data-i18n="view.accrued.freq.1">Annual</option>
                        <option value="2" selected data-i18n="view.accrued.freq.2">Semiannual</option>
                        <option value="4" data-i18n="view.accrued.freq.4">Quarterly</option>
                        <option value="12" data-i18n="view.accrued.freq.12">Monthly</option>
                    </select></label>
                <label><span data-i18n="view.accrued.label.clean">Clean price ($)</span>
                    <input type="number" step="0.01" min="0" name="clean_price" value="980" required></label>
                <label><span data-i18n="view.accrued.label.daycount">Day count</span>
                    <select name="day_count">
                        <option value="thirty360" data-i18n="view.accrued.dc.thirty360">30/360 (corporate/muni)</option>
                        <option value="actual_actual" data-i18n="view.accrued.dc.actualactual">Actual/Actual (Treasury)</option>
                    </select></label>
                <label><span data-i18n="view.accrued.label.last">Last coupon date</span>
                    <input type="date" name="last_coupon" value="2026-01-01" required></label>
                <label><span data-i18n="view.accrued.label.next">Next coupon date</span>
                    <input type="date" name="next_coupon" value="2026-07-01" required></label>
                <label><span data-i18n="view.accrued.label.settle">Settlement date</span>
                    <input type="date" name="settlement" value="2026-04-01" required></label>
            </form>
            <div id="accrued-tools" class="ce-toolbar"></div>
            <button type="button" id="accrued-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="accrued-sens" class="ce-sens"></div>
        </div>
        <div id="accrued-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#accrued-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            face_value: Number(fd.get('face_value')) || 0,
            coupon_rate_pct: Number(fd.get('coupon_rate_pct')) || 0,
            frequency: Number(fd.get('frequency')) || 1,
            clean_price: Number(fd.get('clean_price')) || 0,
            last_coupon: fd.get('last_coupon'),
            next_coupon: fd.get('next_coupon'),
            settlement: fd.get('settlement'),
            day_count: fd.get('day_count'),
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcAccruedInterest(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.accrued.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#accrued-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'accrued-interest.csv' });
    mount.querySelector('#accrued-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['accrued_interest', r.accrued_interest],
        ['dirty_price', r.dirty_price],
        ['clean_price', r.clean_price],
        ['coupon_payment', r.coupon_payment],
        ['days_accrued', r.days_accrued],
        ['accrual_fraction', r.accrual_fraction],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#accrued-result');
    // Line chart: accrued interest as coupon rate sweeps 0 → 12% (linear in coupon).
    const xs = enh.linspace(0, 12, 13);
    const pts = await Promise.all(xs.map(async (c) => {
        const rr = await api.calcAccruedInterest({ ...body, coupon_rate_pct: c });
        return { x: c, y: rr && rr.accrued_interest != null ? rr.accrued_interest : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'coupon %', ylabel: 'accrued $' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.accrued.h2.result">What you pay</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.accrued.card.clean">Clean price</div>
                    <div class="value">${money(r.clean_price)}</div></div>
                <div class="card"><div class="label" data-i18n="view.accrued.card.accrued">Accrued interest</div>
                    <div class="value">${money(r.accrued_interest)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.accrued.card.dirty">Dirty price</div>
                    <div class="value pos">${money(r.dirty_price)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.accrued.row.coupon">Coupon payment</td><td>${money(r.coupon_payment)}</td></tr>
                    <tr><td data-i18n="view.accrued.row.daysaccrued">Days accrued</td><td>${num(r.days_accrued)}</td></tr>
                    <tr><td data-i18n="view.accrued.row.daysperiod">Days in period</td><td>${num(r.days_in_period)}</td></tr>
                    <tr><td data-i18n="view.accrued.row.fraction">Accrual fraction</td><td>${num(r.accrual_fraction)}</td></tr>
                    <tr class="emph"><td data-i18n="view.accrued.row.accrued">Accrued interest</td><td>${money(r.accrued_interest)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#accrued-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: coupon rate 0 → 12%; y: clean price 0.8× → 1.2× current. Output: dirty price.
    const cp = base.clean_price || 980;
    const xVals = enh.linspace(0, 12, 5);
    const yVals = enh.linspace(cp * 0.8, cp * 1.2, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'coupon_rate_pct', yKey: 'clean_price', xVals, yVals, compute: (b) => api.calcAccruedInterest(b), pick: (r) => (r ? r.dirty_price : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + v.toFixed(0)), xfmt: (v) => v.toFixed(1) + '%', yfmt: (v) => '$' + v.toFixed(0), xName: t('view.accrued.label.coupon') || 'Coupon', yName: t('view.accrued.label.clean') || 'Clean' });
}

File: crates/traderview-core/src/accrued_interest.rs
//! Bond accrued interest and dirty (full) price.
//!
//! Between coupon dates a bond's buyer owes the seller the interest that has
//! accrued since the last coupon. The quoted *clean* price excludes it; the
//! *dirty* price the buyer actually pays adds it back:
//!
//! ```text
//! accrued    = coupon_payment × (days_accrued / days_in_period)
//! dirty_price = clean_price + accrued
//! ```
//!
//! Two day-count conventions are supported: 30/360 (US corporate/municipal
//! bond basis) and Actual/Actual (US Treasuries), selected per the security.

use chrono::NaiveDate;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum DayCount {
    /// 30/360 bond basis (corporates, munis).
    Thirty360,
    /// Actual/Actual on the coupon period (Treasuries).
    ActualActual,
}

#[derive(Debug, Clone, Deserialize)]
pub struct AccruedInput {
    /// Par / face value the coupon is computed on (e.g. 1000.0).
    pub face_value: f64,
    /// Annual coupon rate in percent (e.g. 5.0 for a 5% coupon).
    pub coupon_rate_pct: f64,
    /// Coupon payments per year (1, 2, 4, or 12).
    pub frequency: u32,
    /// Quoted price excluding accrued interest, in the same units as face.
    pub clean_price: f64,
    /// Previous coupon date, ISO `YYYY-MM-DD`.
    pub last_coupon: String,
    /// Next coupon date, ISO `YYYY-MM-DD`.
    pub next_coupon: String,
    /// Trade settlement date, ISO `YYYY-MM-DD`.
    pub settlement: String,
    pub day_count: DayCount,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct AccruedResult {
    /// Full coupon paid each period (face × rate / frequency).
    pub coupon_payment: f64,
    /// Days accrued from last coupon to settlement, per the convention.
    pub days_accrued: f64,
    /// Days in the full coupon period, per the convention.
    pub days_in_period: f64,
    /// days_accrued / days_in_period.
    pub accrual_fraction: f64,
    /// Interest accrued since the last coupon.
    pub accrued_interest: f64,
    /// clean_price (echoed for the UI).
    pub clean_price: f64,
    /// clean_price + accrued_interest — what the buyer pays.
    pub dirty_price: f64,
}

/// 30/360 (bond basis) day count between two dates.
fn days_30_360(start: NaiveDate, end: NaiveDate) -> f64 {
    use chrono::Datelike;
    let (y1, m1, mut d1) = (start.year(), start.month() as i32, start.day() as i32);
    let (y2, m2, mut d2) = (end.year(), end.month() as i32, end.day() as i32);
    if d1 == 31 {
        d1 = 30;
    }
    if d2 == 31 && d1 == 30 {
        d2 = 30;
    }
    (360 * (y2 - y1) + 30 * (m2 - m1) + (d2 - d1)) as f64
}

fn parse(s: &str) -> Option<NaiveDate> {
    NaiveDate::parse_from_str(s, "%Y-%m-%d").ok()
}

pub fn analyze(input: &AccruedInput) -> AccruedResult {
    let freq = input.frequency.max(1) as f64;
    let coupon_payment = input.face_value * (input.coupon_rate_pct / 100.0) / freq;

    let last = parse(&input.last_coupon);
    let next = parse(&input.next_coupon);
    let settle = parse(&input.settlement);

    let (days_accrued, days_in_period) = match (last, next, settle) {
        (Some(last), Some(next), Some(settle)) => match input.day_count {
            DayCount::Thirty360 => {
                (days_30_360(last, settle), days_30_360(last, next))
            }
            DayCount::ActualActual => (
                (settle - last).num_days() as f64,
                (next - last).num_days() as f64,
            ),
        },
        _ => (0.0, 0.0),
    };

    let fraction = if days_in_period != 0.0 {
        days_accrued / days_in_period
    } else {
        0.0
    };
    let accrued = coupon_payment * fraction;

    AccruedResult {
        coupon_payment,
        days_accrued,
        days_in_period,
        accrual_fraction: fraction,
        accrued_interest: accrued,
        clean_price: input.clean_price,
        dirty_price: input.clean_price + accrued,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn run(day_count: DayCount, last: &str, next: &str, settle: &str) -> AccruedResult {
        analyze(&AccruedInput {
            face_value: 1000.0,
            coupon_rate_pct: 6.0,
            frequency: 2,
            clean_price: 980.0,
            last_coupon: last.into(),
            next_coupon: next.into(),
            settlement: settle.into(),
            day_count,
        })
    }

    #[test]
    fn coupon_payment_semiannual() {
        // 6% on 1000, twice a year → 30 per coupon.
        let r = run(DayCount::Thirty360, "2026-01-01", "2026-07-01", "2026-01-01");
        assert!(close(r.coupon_payment, 30.0));
    }

    #[test]
    fn thirty_360_half_period() {
        // Jan 1 → Jul 1 is 180 days 30/360; settle Apr 1 is 90 → half.
        let r = run(DayCount::Thirty360, "2026-01-01", "2026-07-01", "2026-04-01");
        assert!(close(r.days_in_period, 180.0));
        assert!(close(r.days_accrued, 90.0));
        assert!(close(r.accrual_fraction, 0.5));
        assert!(close(r.accrued_interest, 15.0));
    }

    #[test]
    fn dirty_is_clean_plus_accrued() {
        let r = run(DayCount::Thirty360, "2026-01-01", "2026-07-01", "2026-04-01");
        assert!(close(r.dirty_price, 980.0 + 15.0));
        assert!(close(r.clean_price, 980.0));
    }

    #[test]
    fn settlement_on_last_coupon_is_zero() {
        let r = run(DayCount::Thirty360, "2026-01-01", "2026-07-01", "2026-01-01");
        assert!(close(r.accrued_interest, 0.0));
        assert!(close(r.dirty_price, 980.0));
    }

    #[test]
    fn settlement_on_next_coupon_is_full() {
        // Just before the next coupon, nearly the whole period has accrued.
        let r = run(DayCount::Thirty360, "2026-01-01", "2026-07-01", "2026-07-01");
        assert!(close(r.accrual_fraction, 1.0));
        assert!(close(r.accrued_interest, 30.0));
    }

    #[test]
    fn actual_actual_uses_calendar_days() {
        // Jan 1 → Jul 1 2026 actual = 181 days; settle Feb 1 = 31 days.
        let r = run(DayCount::ActualActual, "2026-01-01", "2026-07-01", "2026-02-01");
        assert!(close(r.days_in_period, 181.0));
        assert!(close(r.days_accrued, 31.0));
        assert!(close(r.accrued_interest, 30.0 * 31.0 / 181.0));
    }

    #[test]
    fn thirty_360_treats_31st_as_30th() {
        // Jan 31 → settle Feb 28: D1=31→30, so 30/360 = 28 days.
        let r = analyze(&AccruedInput {
            face_value: 1000.0,
            coupon_rate_pct: 6.0,
            frequency: 2,
            clean_price: 1000.0,
            last_coupon: "2026-01-31".into(),
            next_coupon: "2026-07-31".into(),
            settlement: "2026-02-28".into(),
            day_count: DayCount::Thirty360,
        });
        assert!(close(r.days_accrued, 28.0));
        assert!(close(r.days_in_period, 180.0));
    }

    #[test]
    fn bad_dates_guard_to_zero() {
        let r = run(DayCount::Thirty360, "not-a-date", "2026-07-01", "2026-04-01");
        assert!(close(r.accrued_interest, 0.0));
        assert!(close(r.dirty_price, 980.0));
    }
}
